### ISSUE TITLE
Possible Vulnerability - Blind SSRF 

### ISSUE THREAD

**Issue by 0xSpy** [AdguardTeam/AdGuardHome#3493]
While browsing my AdGuard's Administration panel, I noticed that while adding a new allowlists/blocklists (from a custom list to be more precise), there is not security checks on the requested URL, hence the possibility of SSRF. 

HTTP Request: 
```
POST /control/filtering/add_url HTTP/1.1
Host: AdGuard.Host
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:78.0) Gecko/20100101 Firefox/78.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Content-Type: application/json
Content-Length: 57
Connection: close
Cookie: agh_session=session_id

{"url":"http://localhost:22","name":"h","whitelist":true}
```

HTTP Response:
```
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
Date: Mon, 23 Aug 2021 19:37:37 GMT
Content-Length: 181
Connection: close

Couldn't fetch filter from url http://localhost:22: Get "http://localhost:22": net/http: HTTP/1.x transport connection broken: malformed HTTP status code "Raspbian-10+deb10u2+rpt1"
```

In some cases it reveals banners (such as the above HTTP response example), however, this is a blind SSRF.

Because this is a Blind-SSRF, the only impact that I can see is [port scanning. ](https://blog.cobalt.io/from-ssrf-to-port-scanner-3e8ef5921fbf)

Let me know if you need any further clarification.

**Comment by ameshkov** [comment]
Hello, thank you for the report. However, I don't fully understand what's the problem here, could you please explain what exactly could be vulnerable here and, most importantly, how it can be exploited?

> there is not security checks on the requested URL

What sort of security checks can be added?

**Comment by 0xSpy** [comment]
Not fully aware of AdGuard's LLD, so I will give a recommendation from my POV, feel free to neglect my proposal. 

According to my knowledge of the functionality of adding URL that it checks whether or not the requested URL has some text content, if so, it imports it (as a new list). 

To answer your question(s):

> What sort of security checks can be added?

1- A regular expression (ReGeX) that checks if the requested URL is valid (based on your point of view), and basically preventing URL's that are pointing to ports. 
2- Preventing the informative error messages, in this case the messages that shows if the port is open-closed. 
3- Sanatizing the user input in which preventing requests to localhost or 127.0.0.1. 

Please refer to the [OWASP SSRF Prevention](https://cheatsheetseries.owasp.org/cheatsheets/Server_Side_Request_Forgery_Prevention_Cheat_Sheet.html) for a better visibility 


> how it can be exploited?

Well, as I stated in my initial report, since this is a (BLIND) SSRF, only thing that I can see is port scanning.

**Comment by 0xSpy** [comment]
> I am not yet sure this can be considered a vulnerability


Yes, the authentication part could decrease the severity/impact of the vulnerability, but doesn’t change the fact that it’s vulnerable. 

In AppSec, each vulnerability has a score clarifying how serious the vulnerability is.

**Comment by ainar-g** [comment]
I'm not sure if this can be fixed in the general case, because AdGuard Home **is** essentially an HTTP client in that regard.  And also, @ameshkov is right, people who have admin access to AGH can do all sorts of things, including some filesystem access.

And we can't just prohibit **all** private HTTP requests, since there are people who host their filter lists in another machine in their private network (for example, when AGH is used by an organisation), possibly a machine without a proper hostname, or using a localhost proxy of some sort.

We could add a flag, on by default, that would make AGH resolve the host and, if the host is within one of the the locally served ranges, reject this filter.  Same with non-80 and non-443 ports.  But again, people who have access to the admin panel of AGH can do all sorts of things.  Because AGH's admin panel is essentially the same as, say, the one of a router.

**Comment by 0xSpy** [comment]
I can see that you didn't fully understand the vulnerability here. 


In a normal network, users shouldn't access other users internal services, meaning that if a service is running on localhost should only be accessible by the same host. If AdGuard Home is running other services than (HTTP,DNS), exploiting this vulnerability can expose AGH's internal services, so services could be enumerated, in other words, any internal user that has access to AGH's panel can enumerate the machine that is running AGH, this is known as SSRF. 
 

> And we can't just prohibit all private HTTP requests.

You shouldn't. To prevent this attack you should only block requests that directs AdGuard's to add URLs from **localhost**, specially requests that are pointing to ports, this is to prevent:
1- Port Scanning
2- Internal Services Access

> Because AGH's admin panel is essentially the same as, say, the one of a router.

I can relate, but keep in mind that this a WebApp, thus shouldn't be used as [Nmap](https://nmap.org). Also, routers could be vulnerable to various attack, its not something new. 

However, blocking 127.0.0.1 & localhost via ReGeX is not enough, and can be [bypassed](https://book.hacktricks.xyz/pentesting-web/ssrf-server-side-request-forgery#basic-bypass-localhost)


I hope that clarify thing out.

Please let me know if any further clarifications are needed.

**Comment by ainar-g** [comment]
@0xSpy, I have, in fact, specifically mentioned `localhost` in my response:

 >  And we can't just prohibit **all** private HTTP requests, since there are people who host their filter lists in another machine […], **or using a localhost proxy of some sort**.

I agree that this is probably an use case uncommon enough to sacrifice for security, but it's still a breaking change.

 >  However, blocking 127.0.0.1 &amp; localhost via ReGeX is not enough […]

We don't actually need regexps, since we can simply resolve the hosts and check the resulting IPs against all IPv4 and IPv6 forms of `localhost` IPs and null IPs.  The only question is whether to do it as a diableable option or just do it unconditionally.

@ameshkov?

**Comment by ameshkov** [comment]
Sorry, but I don't like the idea of sacrificing any functionality here. If that URL was not protected then this would've been an obvious vulnerability, but this is not the case.

**Comment by 0xSpy** [comment]
> If that URL was **not protected** then this would've been an obvious vulnerability, but this is not the case.

Well, that totally depends on how you define "**protected**", given that there's no password policy enforced (passwords can be only one  character). But maybe securing the app is not a priority right now since the app is running internally at the moment. 

> but I don't like the idea of sacrificing any functionality

I can't seem to understand why sacrificing the functionality if there's another option (patching the vulnerability), would be grateful if you can elaborate.

**Comment by ameshkov** [comment]
> Well, that totally depends on how you define "protected", given that there's no password policy enforced (passwords can be only one character). But maybe securing the app is not a priority right now since the app is running internally at the moment.

Password policy is a good point. Tbh, I'd prefer to increase the overall protection level to fixing issues that can only be exploited if the app is already compromised.

> I can't seem to understand why sacrificing the functionality if there's another option (patching the vulnerability), would be grateful if you can elaborate.

Among proposed solutions, the only one that does not limit functionality in some way is making error messages less detailed. However, I don't like the idea of looking for vulnerabilities in an already compromised app. IMO, the whole app is a vulnerability at this point since it has too much power over the network.

**Comment by 0xSpy** [comment]
2 Questions
1- Do you consider this as a vulnerability? 
2- Any plans to fix it?
